### ISSUE TITLE
Empty codeblock appears when using mention with bold markdown

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18825]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to NewDot
2. Go to any chat with User B
3. First send a message that is bold in nature and which should be within backtick. For example : `*hello*`
4. Notice that the message is sent without any problem
5. Now follow the same process but in this case, mention the user B. For example :`*@userb@gmail.com*`
6. Notice that nothing is displayed but if you try to edit it, you can see the whole text

## Expected Result:
Mentioning user in bold format and within backticks should not behave weirdly and empty message should not be shown
## Actual Result:
Mentioning user in bold format and within backticks behaves weirdly and empty message is shown (works well with the messages , problem is with the mentions only)

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.13
**Reproducible in staging?:** y
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/af1d1166-22ec-416d-9828-0e55d67dc12f

https://github.com/Expensify/App/assets/43996225/05511163-414a-4374-995c-b266615f6051

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683802669770569

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Empty codeblock appears when mention with bold markdown inside codeblock
### What is the root cause of that problem?
Codeblock is rendered in here, the content of text is `props.defaultRendererProps.tnode.data`
https://github.com/Expensify/App/blob/615ebd98c4b32affe2875712187d202a7dc64ab0/src/components/InlineCodeBlock/index.js#L12
https://github.com/Expensify/App/blob/615ebd98c4b32affe2875712187d202a7dc64ab0/src/components/InlineCodeBlock/index.native.js#L17
1. Mention without bold markdown inside codeblock. For example: `@abcde@gmail.com`. The html when is rendered of this message is : "<code><mention-user>@abcde@gmail.com</mention-user></code>". After translation by `react-native-render-html` it is:
```
<TBlock>
  <TText>
    @abcde@gmail.com
  </TText>
</TBlock>
```
The children of codeblock is TText so that `props.defaultRendererProps.tnode.data` is the content that user enter.
2. Mention with bold markdown inside codeblock. For example: `*@abcde@gmail.com*`. The html when is rendered of this message is : "<code>*<mention-user>@abcde@gmail.com</mention-user>*</code>" because bold markdown is disabled inside code markdown that is the rule replace of `expensify-common`. After translation by `react-native-render-html` it is:
```
<TBlock>
  <TPhrasing>
    <TText>
      *
    </TText>
    <TText>
      @abcde@gmail.com
    </TText>
    <TText>
      *
    </TText>
  </TPhrasing>
</TBlock>
```
The children of codeblock is TPhrasing that has threee TText children so that `props.defaultRendererProps.tnode.data` now is undefined, causing this issue.
### What changes do you think we should make in order to solve the problem?
We could change the children of text from `props.defaultRendererProps.tnode.data` to `<TNodeChildrenRenderer tnode={props.defaultRendererProps.tnode} />` to render exactly content for all case.
https://github.com/Expensify/App/blob/615ebd98c4b32affe2875712187d202a7dc64ab0/src/components/InlineCodeBlock/index.js#L12
https://github.com/Expensify/App/blob/615ebd98c4b32affe2875712187d202a7dc64ab0/src/components/InlineCodeBlock/index.native.js#L17

### What alternative solutions did you explore? (Optional)

**Comment by jjcoffee** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Mentions within codeblocks are resulting in an empty codeblock displaying on the FE, despite the message correctly reaching the BE.

### What is the root cause of that problem?
Mentions within codeblocks  are incorrectly being processed by `ExpensiMark`. You can see this in the HTML that is generated when sending the message \`*@userb@gmail.com*\`:

![image](https://github.com/Expensify/App/assets/27287420/644fbe72-c18a-4d16-8971-92afa8dd76c5)

This also happens for *any* additional text included in a codeblock, e.g. \`hello @userb@gmail.com\` will have the same result of an empty codeblock.

If we look at the regex used by the mentions in `ExpensiMark`, we can see that there's only handling for a codeblock with optional whitespace character(s) following it, see [here](https://github.com/Expensify/expensify-common/blob/5f4672c84e19122fc77b4bf1ac092d4919472112/lib/ExpensiMark.js#L75-L83) or below:
```
regex: new RegExp(`((&#x60;|<code>|<pre>)\\s*?)?[\`.a-zA-Z]?@+${CONST.REG_EXP.EMAIL_PART}`, 'gm'),
```

There's an underlying issue that @dukenv0307's [proposal](https://github.com/Expensify/App/issues/18825#issuecomment-1545460748) highlights that is actually causing the codeblock to display as empty, but I don't think it's worth fixing as it's only happening because mentions are being rendered within the codeblock, which shouldn't be happening in the first place!

### What changes do you think we should make in order to solve the problem?
The reason we're capturing the `<code>`, etc. tags in the snippet above is so that we can check with `isValidMention` to ensure the mention isn't surrounded by any disallowed tags (like `<code>`). So we need to update the regex so that they handle codeblocks followed by any character, whilst also maintaining the behaviour that those HTML tags are pulled through for validation.

We can therefore change the regex (for both `userMentions` and `hereMentions`) [here](https://github.com/Expensify/expensify-common/blob/5f4672c84e19122fc77b4bf1ac092d4919472112/lib/ExpensiMark.js#L75-L83) to this:

```
regex: new RegExp(`((&#x60;|<code>|<pre>)(?!.*(&#x60;|<\/code>|<\/pre>).+).*?)?[\`.a-zA-Z]?@+${CONST.REG_EXP.EMAIL_PART}`, 'gm'),
```

Changing the `\s` to the any character (`*`) regex ensures we can have any of those tags followed by any character. 

I've also added a negative lookahead for the same tags as closing tags, so that we don't match other codeblocks within the message, e.g.: `hi` @userb@gmail.com, which would translate into `<code>hi</code> @userb@gmail.com`. That whole string would be matched and pulled into `isValidMention` which would fail as it checks for the validity of the first character in the match, see [here](https://github.com/Expensify/expensify-common/blob/5f4672c84e19122fc77b4bf1ac092d4919472112/lib/str.js#L948-L949)).

### What alternative solutions did you explore? (Optional)
We could also avoid the negative lookahead and use capture groups instead to pass the actual captured mention separately from the context (i.e. any HTML tags). Then in `isValidMention` we can check if the mention itself starts with a valid character, and separately check if it is surrounded by HTML tags.

**Comment by tjferriss** [comment]
I can't get the blank message.

<img width="407" alt="Screenshot 2023-05-15 at 12 10 57 PM" src="https://github.com/Expensify/App/assets/8292562/d184823f-5b22-498f-9509-c844518fba82">

**Comment by dukenv0307** [comment]
@tjferriss you need start after * with "@" before email

**Comment by jjcoffee** [comment]
@tjferriss Just to clarify, the string you need to enter is: `` `*@userb@gmail.com*` `` (with the leading @ so that it's a mention). I am still able to repro on both production and main.

**Comment by kavimuru** [comment]
@tjferriss I am also able to reproduce.

**Comment by jjcoffee** [comment]
Should be getting fixed in #18983
